stdlib/universe/quantile.go

```diff
@@ -258,32 +252,62 @@ func (a *QuantileAgg) NewUIntAgg() execute.DoUIntAgg {
 }
 
 func (a *QuantileAgg) NewFloatAgg() execute.DoFloatAgg {
-	return a.Recycle()
+	q := &QuantileAggState{
+		parent: a,
+	}
+	if len(a.freeDigests) > 0 {
+		q.digest = a.popFreeDigest()
+	} else {
+		a.mem.Account(tdigest.ByteSizeForCompression(a.Compression))
+		q.digest = tdigest.NewWithCompression(a.Compression)
+	}
+	return q
 }
 
 func (a *QuantileAgg) NewStringAgg() execute.DoStringAgg {
 	return nil
 }
 
-func (a *QuantileAgg) DoFloat(vs *array.Float) {
+func (a *QuantileAgg) Dispose() {
+	for i := 0; i < len(a.freeDigests); i++ {
+		a.mem.Account(tdigest.ByteSizeForCompression(a.Compression) * -1)
+	}
+}
+
+type QuantileAggState struct {
+	digest *tdigest.TDigest
+	parent *QuantileAgg
+	ok     bool
+}
+
+func (s *QuantileAggState) DoFloat(vs *array.Float) {
 	for i := 0; i < vs.Len(); i++ {
 		if vs.IsValid(i) {
-			a.digest.Add(vs.Value(i), 1)
-			a.ok = true
+			s.digest.Add(vs.Value(i), 1)
+			s.ok = true
 		}
 	}
 }
 
-func (a *QuantileAgg) Type() flux.ColType {
+func (s *QuantileAggState) Type() flux.ColType {
 	return flux.TFloat
 }
 
-func (a *QuantileAgg) ValueFloat() float64 {
-	return a.digest.Quantile(a.Quantile)
+func (s *QuantileAggState) ValueFloat() float64 {
+	return s.digest.Quantile(s.parent.Quantile)
 }
 
-func (a *QuantileAgg) IsNull() bool {
-	return !a.ok
+func (s *QuantileAggState) IsNull() bool {
+	return !s.ok
+}
+
+func (s *QuantileAggState) Dispose() {
+	if len(s.parent.freeDigests) < cap(s.parent.freeDigests) {
+		s.digest.Reset()
+		s.parent.pushFreeDigest(s.digest)
```

Comment:
You might consider pushing all of this logic into `QuantileAgg` so that all the code that works with `freeDigests` is in methods of the enclosing struct. I think the code would read better that way.